Add desktop wpf article for templates

Thraka opened the pull request:
## Summary

Add new article about templates

Fixes #13471

## Internal previews

- [styles-templates-create-apply-template](https://review.docs.microsoft.com/en-us/dotnet/desktop-wpf/themes/how-to-create-apply-template?branch=pr-en-us-15547)
- [styles-templates-overview#visual-states](https://review.docs.microsoft.com/en-us/dotnet/desktop-wpf/fundamentals/styles-templates-overview?branch=pr-en-us-15547#visual-states)

## Todo

- [x] Convert code to samples repo
- [x] Remove article (and backlinks) that is redirected to the new one.

Thraka committed:
Updated controltemplate section

Thraka committed:
Add template article

Thraka committed:
Add images

Thraka committed:
Add next steps

Thraka committed:
spelling errs

Thraka committed:
temp

Thraka committed:
Merge branch 'master' of github.com:dotnet/docs into adegeo-13471-styles

Thraka committed:
Merge branch 'master' of github.com:dotnet/docs into adegeo-13471-styles

Thraka committed:
Move all code to samples repo

Thraka committed:
Merge branch 'master' of github.com:dotnet/docs into adegeo-13471-styles

Thraka committed:
Del+redirect old article

Thraka committed:
Fix folder for snippets; blarf!

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+# Create a template for a control
+
+With Windows Presentation Foundation (WPF), you can customize an existing control's visual structure and behavior with your own reusable template. Templates can be applied globally to your application, windows and pages, or directly to controls. Most scenarios that require you to create a new control can be covered by instead creating a new template for an existing control.
```
```suggestion
With Windows Presentation Foundation (WPF), you can customize an existing control's visual structure and behavior with your own reusable template. Templates can be applied globally to your application, windows, and pages, or directly to controls. Most scenarios that seemingly require creating a new control can be covered by instead creating a new template for an existing control.
```

Thraka replied:
Would it be more appropriate to keep these as a single "thing" in the list of things? In the context of objects, **App** vs **Windows/Pages** vs **Controls** not **App** vs **Windows** vs **Pages** vs **Controls**

gewarren replied:
I read it as apply:

globally - to app, windows, and pages
directly - to controls

Which is apparently now how I was supposed to read it. So how about:

...globally to your application, controls, or windows and pages.

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+## When to create a ControlTemplate
+
+Controls have many properties, such as <xref:System.Windows.Controls.Border.Background%2A>, <xref:System.Windows.Controls.Control.Foreground%2A>, and <xref:System.Windows.Controls.Control.FontFamily%2A>. These properties control different aspects of the control's appearance, but the changes that you can make by setting these properties are limited. For example, you can set the <xref:System.Window ...
```
```suggestion
Controls have many properties, such as <xref:System.Windows.Controls.Border.Background%2A>, <xref:System.Windows.Controls.Control.Foreground%2A>, and <xref:System.Windows.Controls.Control.FontFamily%2A>. These properties control different aspects of the control's appearance, but the changes that you can make by setting these properties are limited. For example, you can set the <xref:System.Windows.Controls.Control.Foreground%2A> property to blue and <xref:System.Windows.Controls.Control.FontStyle%2A> to italic on a <xref:System.Windows.Controls.CheckBox>. When you want to customize the control's appearance beyond what setting the other properties on the control can do, you create a <xref:System.Windows.Controls.ControlTemplate>.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+Controls have many properties, such as <xref:System.Windows.Controls.Border.Background%2A>, <xref:System.Windows.Controls.Control.Foreground%2A>, and <xref:System.Windows.Controls.Control.FontFamily%2A>. These properties control different aspects of the control's appearance, but the changes that you can make by setting these properties are limited. For example, you can set the <xref:System.Window ...
+
+In most user interfaces, a button has the same general appearance: a rectangle with some text. If you wanted to create a rounded button, you normally have to create a new control that inherits from the button or recreates the functionality of the button. Your new user control would then provide that circular visual.
```
```suggestion
In most user interfaces, a button has the same general appearance: a rectangle with some text. If you wanted to create a rounded button, you could create a new control that inherits from the button or recreates the functionality of the button. In addition, the new user control would provide the circular visual.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+In most user interfaces, a button has the same general appearance: a rectangle with some text. If you wanted to create a rounded button, you normally have to create a new control that inherits from the button or recreates the functionality of the button. Your new user control would then provide that circular visual.
+
+You can avoid creating new controls by customizing the visual layout of an existing control. With a rounded button, you create a <xref:System.Windows.Controls.ControlTemplate> with the wanted visual layout.
```
```suggestion
You can avoid creating new controls by customizing the visual layout of an existing control. With a rounded button, you create a <xref:System.Windows.Controls.ControlTemplate> with the desired visual layout.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+## Prerequisites
+
+Create a new WPF application and in *MainWindow.xaml* (or another window of your choice) set the following properties on the `<Window>` element:
```
```suggestion
Create a new WPF application and, in *MainWindow.xaml* (or another window of your choice), set the following properties on the **Window** element:
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+| **SizetoContent** | `WidthAndHeight` |
+| **MinWidth**      | `250` |
+
+Set the content of the `<Window>` element to the following XAML:
```
I won't correct every occurrence, but style guide says to format XML elements as bold with no angle brackets.

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+|     |     |
+| --- | --- |
+| **x:Key**         | `roundbutton` |
```
Why not pivot the table to have Key and TargetType as column headers, so they'll be automatically bolded?

Thraka replied:
I just think of it as a list, like the table above this one with 3 properties. If you have a big set of property changes, then you can't pivot otherwise you have a long horizontal line of settings.

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+| **x:Key**         | `roundbutton` |
+| **TargetType**    | `Button` |
+
+This control template will be simple. A root element for the control, a <xref:System.Windows.Controls.Grid>. An <xref:System.Windows.Shapes.Ellipse> to draw the rounded appearance of the button. And a <xref:System.Windows.Controls.ContentPresenter> to display the user-specified button content.
```
```suggestion
This control template will be simple:
- a root element for the control, a <xref:System.Windows.Controls.Grid>
- an <xref:System.Windows.Shapes.Ellipse> to draw the rounded appearance of the button
- a <xref:System.Windows.Controls.ContentPresenter> to display the user-specified button content
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+### Ellipse
+
+Notice that the `Fill` and `Stroke` properties of the `<Ellipse>` element are bound to the controls <xref:System.Windows.Controls.Control.Foreground> and <xref:System.Windows.Controls.Control.Background> properties.
```
```suggestion
Notice that the `Fill` and `Stroke` properties of the `<Ellipse>` element are bound to the control's <xref:System.Windows.Controls.Control.Foreground> and <xref:System.Windows.Controls.Control.Background> properties.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+### ContentPresenter
+
+A [\<ContentPresenter>](xref:System.Windows.Controls.ContentPresenter) element is also added to the template. Because this template is designed for a button, we have to take into consideration that the button inherits from <xref:System.Windows.Controls.ContentControl>. The button presents the content of the element. You can set anything inside of the button, such as plain text or even another con ...
```
Style guide also says that link text should not have any special formatting.

gewarren replied:
```suggestion
A [\<ContentPresenter>](xref:System.Windows.Controls.ContentPresenter) element is also added to the template. Because this template is designed for a button, we have to take into consideration that the button inherits from <xref:System.Windows.Controls.ContentControl>. The button presents the content of the element. You can set anything inside of the button, such as plain text or even another control. Both of the following are valid buttons:
```
```suggestion
A [\<ContentPresenter>](xref:System.Windows.Controls.ContentPresenter) element is also added to the template. Because this template is designed for a button, take into consideration that the button inherits from <xref:System.Windows.Controls.ContentControl>. The button presents the content of the element. You can set anything inside of the button, such as plain text or even another control. Both of the following are valid buttons:
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
````diff
@@ -0,0 +1,183 @@
+</Button>
+```
+
+In both of the examples above, the text and the checkbox are set as to the [Button.Content](xref:System.Windows.Controls.ContentControl.Content) property. Whatever is set as the content can be presented through a `<ContentPresenter>`, which is what the template does.
````
```suggestion
In both of the previous examples, the text and the checkbox are set as the [Button.Content](xref:System.Windows.Controls.ContentControl.Content) property. Whatever is set as the content can be presented through a `<ContentPresenter>`, which is what the template does.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+With the dynamic event and property systems that WPF provides, you can watch a specific property for a value and then restyle the template when appropriate. In this example, you'll watch the button's <xref:System.Windows.UIElement.IsMouseOver> property. When the mouse is over the control, style the `<Ellipse>` with a new color. This type of trigger is known as a **PropertyTrigger**.
+
+For this to work though, you'll need to add a name to the `<Ellipse>` that you can reference. Give it the name of `backgroundElement`.
```
```suggestion
For this to work, you'll need to add a name to the `<Ellipse>` that you can reference. Give it the name of **backgroundElement**.
```

Style guide says to format user input as bold.

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+[!code-xaml[EllipseName](~/samples/snippets/desktop-guide/wpf/styles-templates-create-apply-template/csharp/Window4.xaml#EllipseName)]
+
+Next, add a new <xref:System.Windows.Trigger> to the [ControlTemplate.Triggers](xref:System.Windows.Controls.ControlTemplate.Triggers) collection. The trigger will watch the `IsMouseOver` for the value `true`.
```
```suggestion
Next, add a new <xref:System.Windows.Trigger> to the [ControlTemplate.Triggers](xref:System.Windows.Controls.ControlTemplate.Triggers) collection. The trigger will watch the `IsMouseOver` event for the value `true`.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+[!code-xaml[VisualState](~/samples/snippets/desktop-guide/wpf/styles-templates-create-apply-template/csharp/Window6.xaml#VisualState)]
+
+Any animations defined in a `<VisualState>` will be applied when that state is triggered. Create animations for each state. Animations are put inside of a `<Storyboard>` element. For more information about storyboards, see [Storyboards Overview](../../framework/wpf/graphics-multimedia/storyboards-overview.md).
```
```suggestion
Any animations defined in a `<VisualState>` are applied when that state is triggered. Create animations for each state. Animations are put inside of a `<Storyboard>` element. For more information about storyboards, see [Storyboards Overview](../../framework/wpf/graphics-multimedia/storyboards-overview.md).
```

gewarren reviewed (changes requested):
I left some suggestions, but still need to review styles-templates-overview.md. Also, did you run Acrolinx?

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+
+Next, add a `<Setter>` to the `<Trigger>` that changes the `Fill` property of the `<Ellipse>` to a new color.
+
+[!code-xaml[MouseOver](~/samples/snippets/desktop-guide/wpf/styles-templates-create-apply-template/csharp/Window5.xaml#MouseOver)]
```
This sample shows up as empty in the preview.

Thraka replied:
I'm trying to track down why 3 examples show up empty. There are no errors and the references look correct.

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-overview.md:
```diff
@@ -15,7 +15,7 @@ Windows Presentation Foundation (WPF) styling and templating refer to a suite of
 Another feature of the WPF styling model is the separation of presentation and logic. Designers can work on the appearance of an app by using only XAML at the same time that developers work on the programming logic by using C# or Visual Basic.
 
-This overview focuses on the styling and templating aspects of the app and does not discuss any data binding concepts. For information about data binding, see [Data Binding Overview](../data/data-binding-overview.md).
+This overview focuses on the styling and templating aspects of the app and doesn't discuss any data binding concepts. For information about data binding, see [Data Binding Overview](../data/data-binding-overview.md).
```
```suggestion
This overview focuses on the styling and templating aspects of the app and doesn't discuss any data-binding concepts. For information about data binding, see [Data Binding Overview](../data/data-binding-overview.md).
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-overview.md:
```diff
@@ -51,9 +51,58 @@ For more information, see [Create a style for a control](styles-templates-create
+
+Commonly you declare a template as a resource on the `Resources` section of a XAML file. As with all resources, scoping rules apply.
+
+Control templates are a lot more involved than a style. This is more involved because the control template rewrites the visual appearance of the entire control. A style simply applies property changes to the existing control. However, since the template of a control is applied by setting the [Control.Template](xref:System.Windows.Controls.Control.Template) property, you can use a style to define  ...
```
```suggestion
Control templates are a lot more involved than a style. This is because the control template rewrites the visual appearance of the entire control, while a style simply applies property changes to the existing control. However, since the template of a control is applied by setting the [Control.Template](xref:System.Windows.Controls.Control.Template) property, you can use a style to define or set a template.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-overview.md:
```diff
@@ -51,9 +51,58 @@ For more information, see [Create a style for a control](styles-templates-create
+
+Control templates are a lot more involved than a style. This is more involved because the control template rewrites the visual appearance of the entire control. A style simply applies property changes to the existing control. However, since the template of a control is applied by setting the [Control.Template](xref:System.Windows.Controls.Control.Template) property, you can use a style to define  ...
+
+Designers generally allow you to create a copy of an existing template and modify it. For example, in the Visual Studio WPF designer, select a `CheckBox` control, **right-click** > **Edit template** > **Create a copy...**. This command generates a *style that defines a template*.
```
```suggestion
Designers generally allow you to create a copy of an existing template and modify it. For example, in the Visual Studio WPF designer, select a `CheckBox` control, and then right-click and select **Edit template** > **Create a copy**. This command generates a *style that defines a template*.
```

gewarren replied:
Style guide says not to include punctuation in menu options.

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-overview.md:
```diff
@@ -51,9 +51,58 @@ For more information, see [Create a style for a control](styles-templates-create
+
+### TemplateBinding
+
+You may have noticed that the template resource defined above uses the [TemplateBinding Markup Extension](../../framework/wpf/advanced/templatebinding-markup-extension.md). A `TemplateBinding` is an optimized form of a binding for template scenarios, analogous to a binding constructed with `{Binding RelativeSource={RelativeSource TemplatedParent}}`. `TemplateBinding` is useful for binding parts o ...
```
```suggestion
You may have noticed that the template resource defined in the previous section uses the [TemplateBinding Markup Extension](../../framework/wpf/advanced/templatebinding-markup-extension.md). A `TemplateBinding` is an optimized form of a binding for template scenarios, analogous to a binding constructed with `{Binding RelativeSource={RelativeSource TemplatedParent}}`. `TemplateBinding` is useful for binding parts of the template to properties of the control. For example, each control has a <xref:System.Windows.Controls.Control.BorderThickness> property. Use a `TemplateBinding` to manage which element in the template is affected by this control setting.
```

gewarren commented on docs/desktop-wpf/fundamentals/styles-templates-overview.md:
```diff
@@ -118,6 +167,48 @@ In the following illustration, the mouse is pointing to the third item.
 
 In addition to <xref:System.Windows.Trigger> and <xref:System.Windows.EventTrigger>, there are other types of triggers. <xref:System.Windows.MultiTrigger> allows you to set property values based on multiple conditions. You use <xref:System.Windows.DataTrigger> and <xref:System.Windows.MultiDataTrigger> when the property of your condition is data-bound.
 
+## Visual States
```
The C#/VB code samples on line 121 are both showing up, no matter which language you select in the selector.

Thraka commented on docs/desktop-wpf/fundamentals/styles-templates-create-apply-template.md:
```diff
@@ -0,0 +1,183 @@
+---
```
I'm going to follow up with another PR and redo these file names and folder paths.

gewarren replied:
Technically you would need to add redirects then since this link will have gone live.

Thraka replied:
Yes. I wanted to move everything out of here at once. But doing so will make this PR too convoluted. I'm going to move this article now and then the follow up PR will be all the others.

Thraka committed:
Apply suggestions from code review

Thanks! I'll look at the other items soon.

Co-Authored-By: Genevieve Warren <gewarren@microsoft.com>

Thraka committed:
Fix language id on snippet

Thraka committed:
Move article + update toc

Thraka committed:
Move images; fix links

Thraka committed:
Fix xml ref styles

Thraka committed:
fix link

Thraka committed:
fix mdlint

Thraka committed:
Add noloc entries

Thraka committed:
More loc

Thraka committed:
Fix highlight

Thraka committed:
fix redirect

Thraka committed:
Change links from del article to new

Thraka commented:
@gewarren 

1. Got the code snippets fixed so they render
2. Edited the TOC for the deleted file to route to the new file
3. Updated the links to the old file to the new file
4. Renamed the file to remove the two mentions of *template*

gewarren commented:
LGTM!
